server/src/main/java/io/druid/server/router/TieredBrokerHostSelector.java

```diff
@@ -312,11 +313,16 @@ Server pick()
         return null;
       }
 
-      if (roundRobinIndex >= currNodes.size()) {
-        roundRobinIndex %= currNodes.size();
+      int index = roundRobinIndex.get();
+
+      while (true) {
+        int nextIndex = index + 1;
+        if (nextIndex >= currNodes.size()) nextIndex = 0;
+        if (roundRobinIndex.compareAndSet(index, nextIndex)) break;
+        index = roundRobinIndex.get();
       }
 
-      return currNodes.get(roundRobinIndex++);
+      return currNodes.get(index);
```

Comment:
Now this index could be incorrect, because it was set during a pervious call of this method.

Reply:
Oops, my bad. Will fix this.

Reply:
I added `index %= currNodes.size()` back to fix this, is there a better way to do it?

Reply:
I suggest to return to this version: https://github.com/druid-io/druid/pull/5050/commits/ab61f8344af7d67058c531e9d38ff4d5f695616f and just use `nextIndex` instead of `index` below in the code, it's not a problem if indexing will starts from 1 instead of 0. Or the initial value in roundRobinIndex could be set -1.

Reply:
Ok, I wasn't really sure about using `nextIndex` (for the name's sake) but I will revert the change to that and set the initial value of `roundRobinIndex` to -1.